### ISSUE TITLE
How to create a service in dbus-native?

### ISSUE THREAD

**Issue by zhangtianye** [sidorares/node-dbus#88]
I want to use dbus-native to create a dbus server in the backend.
I try to use your example, but after I use your server2.js, I found that I can only see a dbus service named "some.name" in d-feet, but I can't see any object-path and interface in dbus.
Is there some api can I use to create a server?And How can I use your server example?

**Comment by sidorares** [comment]
I can see objects just fine:

![screen shot 2015-07-16 at 8 25 44 pm](https://cloud.githubusercontent.com/assets/173025/8720879/0692c5a4-2bf9-11e5-962f-10e78e94fac5.png)

**Comment by zhangtianye** [comment]
Here is the picture which I meet problem
Is there something different between mint and ubuntu?
![2015-07-16 18 29 55](https://cloud.githubusercontent.com/assets/7447937/8721010/3b278bac-2be9-11e5-8c4a-1862ed8dc46e.png)

**Comment by zhangtianye** [comment]
When I run it in ubuntu, I meet the same problem

**Comment by sidorares** [comment]
hm, this might be something I fixed in master but not pushed to npm. Can you try code from github?

**Comment by zhangtianye** [comment]
Thanks, I can use it now.It makes me feel good.
If I meet some other problem, I will ask you again.

**Comment by sidorares** [comment]
great, I'll publish latest code

**Comment by zhangtianye** [comment]
I changed sessionBus to systemBus,but there is no change in systemBus in d-feet,Is there some ways to do it?

**Comment by zhangtianye** [comment]
I mean that I can't register a systemBus service on dbus. Is there some ways I can do ?

**Comment by zhangtianye** [comment]
I mean that I can't register a systemBus service on dbus. Is there some ways I can do ?

**Comment by sidorares** [comment]
yes, you might need to put config file similar to example here: https://github.com/sidorares/node-dbus#installation

**Comment by sidorares** [comment]
http://dbus.freedesktop.org/doc/dbus-daemon.1.html#configuration_file

**Comment by zhangtianye** [comment]
Do you mean that I need to write a xml to descripe the service Interface ? And then write your dbus-native to descripe method and signals ?

**Comment by sidorares** [comment]
not exactly, this is required to give permission for your service to appear in system bus.

**Comment by zhangtianye** [comment]
![2015-07-20 14 16 52](https://cloud.githubusercontent.com/assets/7447937/8770761/0f45c586-2eea-11e5-83b4-a2a44427c3fb.png)
Here is some problems to use the newest dbus-native read properties. Here is the picture of error and my code.

**Comment by zhangtianye** [comment]
If I use dbus-native write a client to read a property, I get a object like: [ [ { type: 's', child: [] } ], [ '0.9.8.8' ] ],at lib/introspect.js line 137 ,when I read val;
However, when I try to read a property in example server2.js, the val is not a object but a number.At this time, I can noly get '42',not the object like: [ [ { type: 'y', child: [] } ], [42] ].I can't get the object directly from dbus-native,but do something in lib/introspect.js.
Is there some problem exist in dbus-native ?
Can you tell me where is the range of code about how to build a property in dbus-native?
Is there some methods can do for it?

**Comment by sidorares** [comment]
How do you read property from server2.js? Get and GetAll both should return variants according to http://dbus.freedesktop.org/doc/dbus-specification.html#standard-interfaces-properties

**Comment by zhangtianye** [comment]
sessionBus.getService('some.name').getInterface('/com/github/sidorares/1',
	'com.example.service',
	function(err, Interf1){
	if(err){
		console.log('err'+err);
	}else{
		console.log('Interface'+JSON.stringify(Interf1));
		Interf1.TestProperty(function(err, res){
			if(err){
				console.log('err2'+err);
			}else{
				console.log('res'+ res);
			}
		});
	}
});
If I use this code to read the property of server2.js, I'll meet error, and it is in lib/introspect.js about the val ,the type of val is numble, It has no menber we can use as val[0],so it is an error.

**Comment by zhangtianye** [comment]
Here is the code I change it in lib/introspect.js:
 if(typeof val != 'object'){
	 //	console.log(typeof val);
	callback(err, val);
	 //	return;
 }else {
         if (signature.length === 1)
	       callback(err, val[1][0]);
    	  else
    	        callback(err, val[1]);
  }

**Comment by sidorares** [comment]
Closing — this was resolved in the thread at the time, and there is now real documentation for it, which there was not in 2015.

- [Exporting a service](https://github.com/sidorares/dbus-native/blob/master/docs/api.md#exporting-a-service) — the interface descriptor, property access, `PropertiesChanged`
- Runnable examples: [`examples/basic-service.js`](https://github.com/sidorares/dbus-native/blob/master/examples/basic-service.js), [`examples/service-signals.js`](https://github.com/sidorares/dbus-native/blob/master/examples/service-signals.js), [`examples/service/server2.js`](https://github.com/sidorares/dbus-native/blob/master/examples/service/server2.js)

Verified against `dbus-daemon` on 0.11 that the name registers and the object is introspectable:

```
service name registered:   true
introspection lists iface: true
```

Two points from later in this thread that have answers now:

**System bus.** Registering a name on the system bus still needs a policy file in `/etc/dbus-1/system.d/` granting `own` for your name — that is dbus-daemon's permission model, not something a client library can do for you.

**"reading a property gives me `[[{type:'s',child:[]}],['0.9.8.8']]`"** — you were seeing the raw variant. `Properties.Get` returns a variant, so that shape is correct, but it is awkward, and reading it by index is what `variantValue()` is for:

```js
const { variantValue } = require('dbus-native');
const version = variantValue(await props.Get(iface, 'Version'));
```

Through a proxy it is unwrapped for you — `await iface.Version()`. Changing that raw shape to the plain value is #3.
